lib/mongo/server/connection_common.rb

```diff
@@ -52,6 +52,38 @@ def connected?
       # @api private
       attr_reader :pid
 
+      # Build a document that should be used for connection handshake.
+      #
+      # @param [Server::AppMetadata] app_metadata Application metadata
+      #
+      # @return [BSON::Document] Document that should be sent to a server
+      #     for handshake purposes.
+      #
+      # @api private
+      def handshake_document(app_metadata)
+        document = if app_metadata.server_api && app_metadata.server_api[:version]
+                     HELLO_OP_QUERY
+                   else
+                     LEGACY_HELLO_OP_QUERY
```

Comment:
Is this the ismaster payload to be used with op_query (and I gather that the payload differs between op_query and op_msg)? I am wondering if it's better to have a single payload and modify it for the protocol at the point where the protocol message is created.

Reply:
After reading the rest of the diff I think there is abstraction leakage going on where the details of wire protocol ($db field) is leaking into the content documents (i.e. the actual ismaster/hello payloads).

I would prefer to have $db and other fields needed for a particular message type to be added in the same place where the corresponding message class is instantiated (either with or without adding a helper to do so).

Reply:
After a discussion, handshake and check documents building is now properly distributed between layers.